### PR TITLE
Display a notice for expired/expiring subscriptions along with a plugin update message.

### PR DESCRIPTION
**Pull request by manzurahammed**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces the Feature of Showing an expiry subscription message in the plugin table list for WooCommerce extensions that have either expired or are expiring.

### If you haven't configured your development environment for WooCommerce development, please refer to this documentation.
- https://github.com/woocommerce/woocommerce/blob/trunk/README.md#getting-started
- https://github.com/woocommerce/woocommerce/blob/trunk/DEVELOPMENT.md#plugin-development-environments

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Build WooCommerce on your local environment, e.g.: `cd plugins/woocommerce` then `nvm use && pnpm run build --filter=woocommerce`.
2. Connected your store with local woocommerce.com test site. 
3. Download an old version of a Woo Plugins (for example: Checkout Field Editor version 1.7.10). You can find it in the github/woocommerce account or My account > Download page from local woocom test site.
4. Make sure you have active subscription for this plugin.
5. Upload and install the old plugin
6. Make sure you have no expired subscriptions and no subscriptions expiring in the next 7 days
7. Go to the plugin list page and confirm you see no messsage related to subscriptions.
8. Set one of the orders as expiring soon by setting an expiry date in the next 7 days. To do that, run the [following query](https://github.com/Automattic/woocommerce.com/issues/20256#issuecomment-2111773279) and make sure Auto Renew is disabled for this subscription

9. Visit the Woocommerce > Extensions > My subscription page and click the `refresh` button 
10. Now visit the plugin list page again, and should see the notice about the expiring subscription 

Screenshot 2024-05-01 at 1 12 51 PM

11. For expired notice use the following query 

12. Follow the steps 9 and 10 again, and should see this message.

Screenshot 2024-05-01 at 1 25 30 PM

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by hesahesa** [comment]
@manzurahammed will we add track events later in different PR or should we add in this PR?

**Comment by manzurahammed** [comment]
> @manzurahammed will we add track events later in different PR or should we add in this PR?

Plan to do this on a separate PR

**Comment by gedex** [comment]
> Plan to do this on a separate PR

I'd suggest to add any related Tracks events in the same PR to avoid missing the release

**Comment by hesahesa** [comment]
> I'd suggest to add any related Tracks events in the same PR

I'll assist in adding Tracks events for this PR

update: added in https://github.com/woocommerce/woocommerce/pull/47076/commits/b04f161fb43763a06a2f2cc549364b18caecf0ea

**Review by hesahesa — APPROVED** [review]
LGTM, can someone else also double check the Track events that I committed?

**Comment by gedex** [comment]
#### My tests

- [x] site is not connected; no message introduced by this PR
- [x] no expired or expiring; no message
- [x] expiring and autorenew disabled; enable auto-renew message
- [x] expired; renew (with price) message
- [x] renew link Tracks event (`woo_enable_autorenew_in_plugins_clicked`)
- [x] enable auto-renew Tracks event (`wcadmin_woo_renew_subscription_in_plugins_clicked`)

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php [hunk 1]
```diff
@@ -231,6 +243,64 @@ public static function add_install_marketplace_plugin_message( $plugin_data, $re
...
+		}
+
+		// Prepare the expiry notice based on subscription status.
+		$expiry_notice = '';
+		if ( ! empty( $subscription['expired'] ) && ! $subscription['lifetime'] ) {
+			$expiry_notice = sprintf(
+			/* translators: 1: URL to My Subscriptions page 2: Product price */
+				__( ' Your subscription expired, <a href="%1$s" class="woocommerce-renew-subscription">renew for %2$s</a> to update.', 'woocommerce' ),
+				'https://woocommerce.com/my-account/my-subscriptions/',
+				$subscription['product_regular_price']
```
Comment (gedex) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: can we do a check to make sure `$subscription['product_regular_price']` is not empty? I got notice rendered due to that.

image

Comment (gedex) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: `WC_Helper::get_subscriptions()` is likely won't have `product_regular_price` prop set before hitting the latest endpoint resp

Comment (hesahesa) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: huh interesting, so in the case of currently cached response in the store they don't have the value cached before next API call

Comment (gedex) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: When `$subscription['product_regular_price']` is not set, we can do a generic messaging:

```
Your subscription expired, renew to update.
```

Comment (manzurahammed) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: Nice catch @gedex 
Yes, We can change the message when `product_regular_price ` is not set.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php [hunk 2]
```diff
@@ -231,6 +243,64 @@ public static function add_install_marketplace_plugin_message( $plugin_data, $re
...
+			return;
+		}
+
+		// Prepare the expiry notice based on subscription status.
+		$expiry_notice = '';
+		if ( ! empty( $subscription['expired'] ) && ! $subscription['lifetime'] ) {
+			$expiry_notice = sprintf(
+			/* translators: 1: URL to My Subscriptions page 2: Product price */
+				__( ' Your subscription expired, <a href="%1$s" class="woocommerce-renew-subscription">renew for %2$s</a> to update.', 'woocommerce' ),
+				'https://woocommerce.com/my-account/my-subscriptions/',
```
Comment (gedex) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: missing esc_url ?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php (hunk 1 above)
Comment (gedex) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: also, missing esc_html there

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php (hunk 2 above)
Comment (gedex) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: nvm, it's a literal string

**Review by gedex — APPROVED** [review]
Overall LGTM. Tested and works as expected.

Please add the regular price check to fix the notice being rendered.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php (hunk 1 above)
Comment (manzurahammed) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: Addressed [0a5eecb](https://github.com/woocommerce/woocommerce/pull/47076/commits/0a5eecbb393d7f11b90465275f614f205bda3f3f)

**Pull request merged**